### PR TITLE
storage: remove custom operator tokens and legacy async integration

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

This is the final installment of switching all async operators into the new async integration that solves a lot of the footguns that were present previously.

This PR also replaces all custom tokens of manually wired up `oneshot` channels to the token provided by `AsyncOperatorBuilder`.

### Tips for reviewer

* The diff is much smaller if viewed with whitespace hidden.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/storage/src/render/persist_sink.rs [hunk 1]
```diff
@@ -195,99 +162,102 @@ pub fn render<G>(
                         }
                     }
                 }
+                Event::Progress(input_upper) => {
```
Comment (guswynn) [inline src/storage/src/render/persist_sink.rs:165]: this reminds me, does timely periodically send an extra activation for just frontier updates, where no data will be in the input buffers? of could this branch be starved?

File: src/storage/src/sink/kafka.rs [hunk 2]
```diff
@@ -1016,45 +1014,40 @@ where
...
-    let shutdown_flush = Cell::new(false);
-
     // We want exactly one worker to send all the data to the sink topic.
     let hashed_id = id.hashed();
     let is_active_worker = (hashed_id as usize) % scope.peers() == scope.index();
 
     let mut input = builder.new_input(&stream, Exchange::new(move |_| hashed_id));
 
-    builder.build(move |_capabilities| async move {
+    let button = builder.build(move |_capabilities| async move {
```
Comment (guswynn) [inline src/storage/src/sink/kafka.rs:1023]: wait i feel like we should explicitly drop these

File: src/storage/src/sink/kafka.rs [hunk 3]
```diff
@@ -1091,195 +1084,184 @@ where
...
-
-                    // Approximately one last attempt to push anything pending to kafka before closing.
-                    if !shutdown_flush.get() {
-                        debug!("flushing kafka producer for sink: {}", &s.name);
-                        let _ = s.producer.flush().await;
-                        shutdown_flush.set(true);
-                    }
+        while let Some(event) = input.next().await {
+            match event {
+                Event::Data(_, rows) => {
```
Comment (guswynn) [inline src/storage/src/sink/kafka.rs:1089]: so much better

File: src/storage/src/sink/kafka.rs [hunk 4]
```diff
@@ -1091,195 +1084,184 @@ where
...
-            tokio::select! {
-                _ = shutdown_rx.as_mut() => {
-                    debug!("shutting down sink: {}", &s.name);
-
-                    // Approximately one last attempt to push anything pending to kafka before closing.
-                    if !shutdown_flush.get() {
-                        debug!("flushing kafka producer for sink: {}", &s.name);
-                        let _ = s.producer.flush().await;
-                        shutdown_flush.set(true);
-                    }
```
Comment (guswynn) [inline src/storage/src/sink/kafka.rs:1105]: this was just best-effort right? so we are just removing it?

File: src/storage/src/source/resumption.rs [hunk 5]
```diff
@@ -88,33 +87,24 @@ where
...
-                            "resumption({id}) {worker_id}/{worker_count}: calculated \
-                            new resumption frontier",
-                        );
-
-                        cap_set.downgrade(&*new_upper);
-                        upper = new_upper;
-                    }
-                }
-                futures::future::Either::Right(_) => return,
+            interval.tick().await;
```
Comment (guswynn) [inline src/storage/src/source/resumption.rs:91]: dank

File: src/storage/src/source/source_reader_pipeline.rs [hunk 6]
```diff
@@ -1114,17 +1102,10 @@ where
...
-        tokio::pin!(shutdown_rx);
         let mut input_frontier = Antichain::from_elem(Timestamp::default());
         loop {
             // AsyncInputHandle::next is cancel safe
             tokio::select! {
-                biased;
-                // Make sure we don't accidentally mint new updates when this source has
-                // been dropped. This way, we also make sure to not react to spurious
-                // frontier advancements to `[]` that happen when the input source operator
-                // is shutting down.
```
Comment (guswynn) [inline src/storage/src/source/source_reader_pipeline.rs:1126]: we ensure this by drop being implemented normally with the shutdown button right?

File: src/timely-util/src/activator.rs [hunk 7]
```diff
@@ -225,30 +224,3 @@ impl ArcActivatorInner {
         self.activated = 0;
     }
 }
-
-/// An activator that ignores new activations if one is already pending.
```
Comment (guswynn) [inline src/timely-util/src/activator.rs:229]: why did this even exist lol

**Review by guswynn — APPROVED** [review]
I'm shook that you were able to do this in one go; CI should be able to test this fully!

I have a few nits and inline questions, but for the most part seems good!

(for other reviewers, hiding whitespace changes is your friend here!

File: src/storage/src/render/persist_sink.rs [hunk 8]
```diff
@@ -99,66 +92,40 @@ pub fn render<G>(
         current_upper.borrow_mut().clear();
     }
 
-    let weak_token = Rc::downgrade(&token);
-
     let persist_clients = Arc::clone(&storage_state.persist_clients);
-    persist_op.build_async(
-        scope.clone(),
-        move |mut capabilities, frontiers, scheduler| async move {
-            capabilities.clear();
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/render/persist_sink.rs:108]: The `AsyncOperatorBuilder` follows the same pattern of the normal builder. `capabilities` are passed to the constructor closure but if you don't capture it in your returned future it is dropped immediately so we don't need to clear it

File: src/storage/src/render/persist_sink.rs [hunk 9]
```diff
@@ -99,66 +92,40 @@ pub fn render<G>(
...
-                write.maybe_heartbeat_writer().await;
-                trace!(
-                    "storage persist_sink {}/{}: writer heartbeating write handle",
-                    src_id,
-                    metadata.data_shard
-                );
-                // NOTE: We schedule ourselves to make sure that we keep
-                // heartbeating even in cases where there are no changes in our
-                // inputs (updates or frontier changes).
-                activator.activate_after(Duration::from_secs(60));
```
[the remark this answers is not in the document]
Comment (petrosagg) [reply src/storage/src/render/persist_sink.rs:159]: No, persist handles auto-heartbeat now

File: src/storage/src/render/persist_sink.rs (hunk 1 above)
Comment (petrosagg) [reply src/storage/src/render/persist_sink.rs:165]: It is [configurable](https://docs.rs/timely/latest/timely/dataflow/operators/generic/builder_rc/struct.OperatorBuilder.html#method.set_notify) and [true by default](https://github.com/TimelyDataflow/timely-dataflow/blob/master/timely/src/dataflow/operators/generic/builder_raw.rs#L37)

File: src/storage/src/sink/kafka.rs (hunk 2 above)
Comment (petrosagg) [reply src/storage/src/sink/kafka.rs:1023]: you mean the button?

File: src/storage/src/source/source_reader_pipeline.rs (hunk 6 above)
Comment (petrosagg) [reply src/storage/src/source/source_reader_pipeline.rs:1126]: Yes, the whole future is dropped when the button is dropped so no more work will happen

File: src/storage/src/sink/kafka.rs (hunk 4 above)
Comment (petrosagg) [reply src/storage/src/sink/kafka.rs:1105]: exactly

File: src/storage/src/sink/kafka.rs (hunk 2 above)
Comment (guswynn) [reply src/storage/src/sink/kafka.rs:1023]: the caps!

**Pull request merged**